cmd/ndt7-client/internal/emitter/batch.go

```diff
@@ -35,44 +35,44 @@ func (b Batch) emitInterface(any interface{}) error {
 }
 
 type batchEvent struct {
-	Key   string      `json:"key"`
-	Value interface{} `json:"value"`
+	Key   string
+	Value interface{}
```

Comment:
should be `batchValue` not `interface{}`, right?

Reply:
No, because there we can also put a Measurement.

Reply:
As in, I changed to `batchValue` and got:

```
$ go build -v ./cmd/ndt7-client/
github.com/m-lab/ndt7-client-go/cmd/ndt7-client/internal/emitter
# github.com/m-lab/ndt7-client-go/cmd/ndt7-client/internal/emitter
cmd/ndt7-client/internal/emitter/batch.go:84:3: cannot use m (type *spec.Measurement) as type batchValue in field value
cmd/ndt7-client/internal/emitter/batch.go:92:3: cannot use m (type *spec.Measurement) as type batchValue in field value
```

Reply:
In theory we can do this:

```diff
diff --git a/cmd/ndt7-client/internal/emitter/batch.go b/cmd/ndt7-client/internal/emitter/batch.go
index 1f2a256..5cabf0e 100644
--- a/cmd/ndt7-client/internal/emitter/batch.go
+++ b/cmd/ndt7-client/internal/emitter/batch.go
@@ -36,13 +36,14 @@ func (b Batch) emitInterface(any interface{}) error {
 
 type batchEvent struct {
 	Key   string
-	Value interface{}
+	Value batchValue
 }
 
 type batchValue struct {
-	Failure string
-	Server  string
-	Test    string
+	*spec.Measurement
+	Failure string `json:",omitempty"`
+	Server  string `json:",omitempty"`
+	Test    string `json:",omitempty"`
 }
 
 // OnStarting emits the starting event
@@ -81,7 +82,9 @@ func (b Batch) OnConnected(test, fqdn string) error {
 func (b Batch) OnDownloadEvent(m *spec.Measurement) error {
 	return b.emitInterface(batchEvent{
 		Key:   "measurement",
-		Value: m,
+		Value: batchValue{
+			Measurement: m,
+		},
 	})
 }
 
@@ -89,7 +92,9 @@ func (b Batch) OnDownloadEvent(m *spec.Measurement) error {
 func (b Batch) OnUploadEvent(m *spec.Measurement) error {
 	return b.emitInterface(batchEvent{
 		Key:   "measurement",
-		Value: m,
+		Value: batchValue{
+			Measurement: m,
+		},
 	})
 }
 
```

is it worth it?

Reply:
It is worth it.

`interface{}` and casting is a terrible pattern and we should strive to avoid it whenever possible, for the same reason that we avoid passing around `void*` in C/C++ or `Object` in Java.

Reply:
Right, then I've made it working! Please, take another look!